app/src/main/java/org/alphatilesapps/alphatiles/Validator.java

```diff
@@ -0,0 +1,992 @@
+            Tab keyboardTab = langPackGoogleSheet.getTabFromName("keyboard");
+
+            ArrayList<String> themeColorCol = keyboardTab.getCol(1);
+            themeColorCol.removeAll(Set.of("0", "1", "2", "3", "4"));
+            if (themeColorCol.size() > 0){
+                warnings.add("colum 2 of keyboard should only have numbers 0-4");
+            }
+
+            // the below section compares they keys in keyboard to the words in wordlist
+            if (!keyboardTab.getCol(0).contains(".")) {
+                keyboardTab.add(new ArrayList<>(List.of(".", "0")));
+            }
+
+            // this is a map of each key to how many times it is used in the wordslist
+            Map<String, Integer> keyUsage = new HashMap<>();
+            ArrayList<String> keysList = keyboardTab.getCol(0);
+            for (String key : keysList) {
+                keyUsage.put(key, 0);
+            }
+            for (String cell : langPackGoogleSheet.getTabFromName("wordlist").getCol(1)) {
+                for (String letter : cell.split("")) {
+                    if (!keyUsage.containsKey(letter)) {
+                        warnings.add("In wordList, the word " + cell + " contains the letter " + letter +
+                                " which is not in the keyboard.");
+                    } else {
+                        keyUsage.put(letter, keyUsage.get(letter) + 1);
+                    }
+                }
+            }
+            for (Map.Entry<String, Integer> entry : keyUsage.entrySet()) {
+                if (entry.getValue() < 6) {
+                    warnings.add("In wordList.txt, the letter " + entry.getKey() + " is only used" +
+                            " in " + entry.getValue() + " words. It is recommended that each letter be" +
+                            " used in at least 6 words.");
+                }
+            }
+        } catch (ValidatorException e) {
+            warnings.add(GENERIC_WARNING);
+        }
+
+        //todo is checking headers still a necessary check for gameTiles or others?.. aslo non initial letters black box error?
+        try {
+            // the below section compares the tiles in gameTiles to the words in wordlist
+            Map<String, Integer> tileUsage = new HashMap<>();
+            int longWords = 0;
+            tileUsage.put(".", 0);
+            tileUsage.put("'", 0);
+            for (String tile : langPackGoogleSheet.getTabFromName("gametiles").getCol(0)) {
+                tileUsage.put(tile, 0);
+            }
+            for (String word : langPackGoogleSheet.getTabFromName("wordlist").getCol(1)) {
+                // the below numTilesInWord method parses a word into the tiles in the tileUsage dictionary,
+                // returning the number of tiles in the provided word.
+                // it also updates the tileUsage dictionary so the dictionary counts how many times each tile appears
+                int tileCounter = numTilesInWord(word, tileUsage);
```

Comment:
There is more complex parsing being done in certain non-Roman scripts (Thai and Lao so far). This is captured in Alpha Tiles code in Start of the thai-script-tile-parsing branch, which assumes an existing language pack. Eventually, we'll need parsing like the parsing in Start.tileList.parseWordIntoTiles(), including helper methods.